README
-- Basic usage
$dbp = new DbPatcher(
    <database-config>,
    <patch-directory>
);
$dbp->run();


CHANGELOG
CHANGELOG
=========
  * 2012-11-20 - DbPatcher 1.0
    - First release


DbPatcher.php
<?php
/**
 * This source file is part of DbPatcher.
 *
 * DbPatcher is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * DbPatcher is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License along
 * with DbPatcher. If not, see <http://www.gnu.org/licenses/gpl-3.0.html>.
 *
 * PHP version 5.3
 *
 * @category DbPatcher
 * @package  DbPatcher
 * @author   Sliim <sliim@mailoo.org>
 * @license  GNU/GPL http://www.gnu.org/licenses/gpl-3.0.html
 * @link     http://www.sliim-projects.eu
 */

require_once 'library/DbPatcher/Commander.php';
require_once 'library/DbPatcher/Prompt.php';
require_once 'library/DbPatcher/Patch.php';
require_once 'library/DbPatcher/Patcher.php';

use DbPatcher\Patcher;
use DbPatcher\Commander;
use DbPatcher\Prompt;

/**
 * DbPatcher
 *
 * @category DbPatcher
 * @package  DbPatcher
 */
class DbPatcher
{

    /**
     * @var PDO
     */
    private $connection = null;

    /**
     * @var DbPatcher\Patcher
     */
    private $patcher = null;

    /**
     * @var DbPatcher\Prompt
     */
    private $prompt = null;

    /**
     * @var DbPatcher\Commander
     */
    private $commander = null;

    /**
     * DbPatcher contructor
     *
     * @param array  $dbConfig Database configuration
     * @param string $patchDir Patch directory
     *
     * @throws RuntimeException
     *
     * @return void
     */
    public function __construct(array $dbConfig, $patchDir)
    {
        $diff = array_diff(
            array('driver', 'hostname', 'username', 'password'),
            array_keys($dbConfig)
        );

        if (!empty($diff)) {
            throw new RuntimeException(
                'Missing (' . implode(', ', $diff) . ') in database configuration'
            );
        }

        if (!file_exists($patchDir)) {
            throw new RuntimeException('Patch directory does not exists!');
        }

        $this->connection = new PDO(
            $dbConfig['driver'] . ':host=' .
            $dbConfig['hostname'] . ';dbname=' .
            $dbConfig['dbname'],
            $dbConfig['username'],
            $dbConfig['password']
        );

        $this->connection->setAttribute(PDO::ATTR_ERRMODE, PDO::ERRMODE_EXCEPTION);
        $this->patchDir = $patchDir;
    }

    /**
     * Run DbPatcher
     *
     * @return void
     */
    public function run()
    {
        $this->getPrompt()->run();
    }

    /**
     * Patcher getter
     *
     * @return Patcher
     */
    public function getPatcher()
    {
        if (is_null($this->patcher)) {
            $this->patcher = new Patcher($this->patchDir, $this->connection);
        }

        return $this->patcher;
    }

    /**
     * Commander getter
     *
     * @return Commander
     */
    public function getCommander()
    {
        if (is_null($this->commander)) {
            $this->commander = new Commander();
            $this->commander
                ->addCommand('quit', array($this->commander, 'cExit'), 'Alias of `exit`')
                ->addCommand('list', array($this, 'getList'), 'Patch list')
                ->addCommand('status', array($this, 'status'), 'Database status')
                ->addCommand('up', array($this->getPatcher(), 'upgrade'), 'Upgrade database')
                ->addCommand('down', array($this->getPatcher(), 'downgrade'), 'Downgrade database');
        }

        return $this->commander;
    }

    /**
     * Prompt getter
     *
     * @return Prompt
     */
    public function getPrompt()
    {
        if (is_null($this->prompt)) {
            $this->prompt = new Prompt($this->getCommander());
            $this->prompt->setPrompt('dbpatcher>>>');
        }

        return $this->prompt;
    }

    /**
     * Patcher setter
     *
     * @param Patcher $patcher Patcher to set
     *
     * @return DbPatcher
     */
    public function setPatcher(Patcher $patcher)
    {
        $this->patcher = $patcher;
        return $this;
    }

    /**
     * Commander setter
     *
     * @param Commander $commander Commander to set
     *
     * @return DbPatcher
     */
    public function setCommander(Commander $commander)
    {
        $this->commander = $commander;
        return $this;
    }

    /**
     * Prompt setter
     *
     * @param Prompt $prompt Prompt to set
     *
     * @return DbPatcher
     */
    public function setPrompt(Prompt $prompt)
    {
        $this->prompt = $prompt;
        return $this;
    }

    /**
     * Print database status
     *
     * @return void
     */
    public function status()
    {
        $status = $this->getPatcher()->status();

        echo 'Database status: version ' . $status['status'] . PHP_EOL;

        if (empty($status['toInstall'])) {
            echo 'Database up-to-date!' . PHP_EOL;
        } else {
            echo 'Patch to install:' . PHP_EOL;
            foreach ($status['toInstall'] as $patch) {
                echo "\t-$patch\r\n";
            }
        }

        echo PHP_EOL;
    }

    /**
     * Print patch list
     *
     * @return void
     */
    public function getList()
    {
        $list = $this->getPatcher()->getList();

        echo 'Patch list:' . PHP_EOL;
        foreach ($list as $patch) {
            echo "\t-$patch\r\n";
        }

        echo PHP_EOL;
    }
}


config/database.php
<?php
/**
 * This source file is part of DbPatcher.
 *
 * DbPatcher is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * DbPatcher is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License along
 * with DbPatcher. If not, see <http://www.gnu.org/licenses/gpl-3.0.html>.
 *
 * PHP version 5.3
 *
 * @category DbPatcher
 * @package  Configuration
 * @author   Sliim <sliim@mailoo.org>
 * @license  GNU/GPL http://www.gnu.org/licenses/gpl-3.0.html
 * @link     http://www.sliim-projects.eu
 */

return array(
    'driver'   => 'pgsql',
    'hostname' => '...',
    'username' => '...',
    'password' => '...',
    'dbname'   => '...',
);


library/DbPatcher/Patch.php
<?php
/**
 * This source file is part of DbPatcher.
 *
 * DbPatcher is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * DbPatcher is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License along
 * with DbPatcher. If not, see <http://www.gnu.org/licenses/gpl-3.0.html>.
 *
 * PHP version 5.3
 *
 * @category DbPatcher
 * @package  Patch
 * @author   Sliim <sliim@mailoo.org>
 * @license  GNU/GPL http://www.gnu.org/licenses/gpl-3.0.html
 * @link     http://www.sliim-projects.eu
 */

namespace DbPatcher;

/**
 * Abstract Patch
 *
 * @category DbPatcher
 * @package  Patch
 */
abstract class Patch
{

    /**
     * @var PDO
     */
    private $connection;

    /**
     * @var array
     */
    private $queries = array();

    /**
     * Constructor, set connection
     *
     * @param PDO $con Connexion
     *
     * @return void
     */
    public function __construct(\PDO $con)
    {
        $this->connection = $con;
    }

    /**
     * Upgrade method
     *
     * @return void
     */
    abstract public function upgrade();

    /**
     * Downgrade method
     *
     * @return void
     */
    abstract public function downgrade();

    /**
     * Prepare a query
     *
     * @param string $sql Query to prepare
     *
     * @return void
     */
    final protected function prepare($sql)
    {
        echo '[INFO] Prepare query: ' . $sql . PHP_EOL;
        $this->queries[] = $this->connection->prepare($sql);
    }

    /**
     * Execute all prepared queries
     *
     * @throws \RuntimeException
     *
     * @return void
     */
    final public function execute()
    {
        $this->connection->beginTransaction();
        foreach ($this->queries as $query) {
            echo '[INFO] Execute query: ' . $query->queryString . PHP_EOL;
            try {
                $query->execute();
            } catch (\PDOException $e) {
                $this->connection->rollBack();
                throw new \RuntimeException(
                    'Error on query `' . $query->queryString .
                    '`.. Process to Rollback!' . PHP_EOL . 'Reason: ' .
                    $e->getMessage()
                );
            }
        }

        $this->connection->commit();
    }
}


library/DbPatcher/Prompt.php
<?php
/**
 * This source file is part of DbPatcher.
 *
 * DbPatcher is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * DbPatcher is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License along
 * with DbPatcher. If not, see <http://www.gnu.org/licenses/gpl-3.0.html>.
 *
 * PHP version 5.3
 *
 * @category DbPatcher
 * @package  Prompt
 * @author   Sliim <sliim@mailoo.org>
 * @license  GNU/GPL http://www.gnu.org/licenses/gpl-3.0.html
 * @link     http://www.sliim-projects.eu
 */

namespace DbPatcher;

/**
 * Prompt class
 *
 * @category DbPatcher
 * @package  Prompt
 *
 * @todo implement history
 */
class Prompt
{

    /**
     * @var Commander
     */
    private $commander = null;

    /**
     * @var string
     */
    private $prompt = '=>>';

    /**
     * Constructor
     * Set commander
     *
     * @param Commander $commander Commander object
     *
     * @return void
     */
    public function __construct(Commander $commander)
    {
        $this->commander = $commander;
    }

    /**
     * Set prompt string
     *
     * @param string $prompt Prompt to set
     *
     * @return void
     */
    public function setPrompt($prompt)
    {
        $this->prompt = trim($prompt);
    }

    /**
     * Run script, read STDIN on infinite loop
     *
     * @return void
     */
    public function run()
    {
        $c = '';
        do {
            try {
                $c = trim($c);
                if (!empty($c)) {
                    $args   = explode(' ', $c);
                    $c      = array_shift($args);
                    $result = call_user_func_array(array($this->commander, $c), $args);

                    switch (gettype($result)) {
                        case 'string':
                        case 'integer':
                        case 'double':
                            echo $result . PHP_EOL;
                            break;
                        case 'NULL':
                            break;
                        default:
                            var_dump($result);
                            break;
                    }
                }
            } catch (\Exception $e) {
                if ($e->getCode() === Commander::EXIT_CODE) {
                    echo 'exiting' . PHP_EOL;
                    return true;
                } else {
                    echo "\033[31m[" . get_class($e) . "]\033[0m " . $e->getMessage() . PHP_EOL;
                    echo 'File: ' . $e->getFile() . ':' . $e->getLine() . PHP_EOL;
                }
            }
            echo $this->prompt . ' ';
        } while ($c = fgets(STDIN));
    }
}


library/DbPatcher/Commander.php
<?php
/**
 * This source file is part of DbPatcher.
 *
 * DbPatcher is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * DbPatcher is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License along
 * with DbPatcher. If not, see <http://www.gnu.org/licenses/gpl-3.0.html>.
 *
 * PHP version 5.3
 *
 * @category DbPatcher
 * @package  Commander
 * @author   Sliim <sliim@mailoo.org>
 * @license  GNU/GPL http://www.gnu.org/licenses/gpl-3.0.html
 * @link     http://www.sliim-projects.eu
 */

namespace DbPatcher;

/**
 * Command manager
 *
 * @category DbPatcher
 * @package  Commander
 */
class Commander
{

    /** @var int */
    const EXIT_CODE = 1337;

    /**
     * Commands list
     * @var array
     */
    private $commands = array();


    /**
     * Constructor
     *
     * @throws \RuntimeException if sapi isn't cli
     *
     * @return void
     */
    public function __construct()
    {
        if (php_sapi_name() !== 'cli') {
            throw new \RuntimeException('Class usable in CLI only !');
        }
    }

    /**
     * Magic call
     * Execute a command if exists
     *
     * @param string $command Command to execute
     * @param array  $args    Command's arguments
     *
     * @throws \BadMethodCallException When command not found
     *
     * @return mixed Command's result
     */
    final public function __call($command, array $args)
    {
        $method = 'c' . ucfirst($command);
        if (method_exists($this, $method)) {
            return call_user_func_array(array($this, $method), $args);
        }

        if (array_key_exists($command, $this->commands)) {
            return call_user_func_array($this->commands[$command]['command'], $args);
        }

        throw new \BadMethodCallException('Command `' . $command . '` not found.');
    }

    /**
     * Add a command
     *
     * @param string $name        Command's name
     * @param mixed  $command     String for a function name or array(object => method)
     * @param string $description Command's description
     *
     * @throws \BadMethodCallException
     * @throws \BadFunctionCallException
     *
     * @return Commander
     */
    public function addCommand($name, $command, $description = '')
    {
        if (is_array($command)) {
            if (!method_exists($command[0], $command[1])) {
                throw new \BadMethodCallException(
                    'Method `' . $method . '` not found in ' . get_class($object)
                );
            }
        } else {
            if (!function_exists($command)) {
                throw new \BadFunctionCallException('Function `' . $command . '` doesn\'t exists');
            }
        }

        $this->commands[$name] = array(
            'command'     => $command,
            'description' => $description,
        );
        return $this;
    }

    /**
     * Help command
     *
     * @todo Get possible arguments for each command
     *
     * @return string
     */
    public function cHelp()
    {
        $help  = 'Commands:' . PHP_EOL;
        $help .= '--------' . PHP_EOL;
        foreach ($this->commands as $name => $command) {
            $help .= $name . "\t=> " . $command['description'] . PHP_EOL;
        }

        $help .= PHP_EOL . '------' . PHP_EOL;
        $help .= "help\t=> Show this help" . PHP_EOL;
        $help .= "exit\t=> Exit script" . PHP_EOL;

        return $help;
    }

    /**
     * Exit command
     *
     * @throws \Exception with EXIT_CODE constant
     *
     * @return void
     */
    public function cExit()
    {
        throw new \Exception('exit', static::EXIT_CODE);
    }
}


library/DbPatcher/Patcher.php
<?php
/**
 * This source file is part of DbPatcher.
 *
 * DbPatcher is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * DbPatcher is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License along
 * with DbPatcher. If not, see <http://www.gnu.org/licenses/gpl-3.0.html>.
 *
 * PHP version 5.3
 *
 * @category DbPatcher
 * @package  Patcher
 * @author   Sliim <sliim@mailoo.org>
 * @license  GNU/GPL http://www.gnu.org/licenses/gpl-3.0.html
 * @link     http://www.sliim-projects.eu
 */

namespace DbPatcher;

/**
 * Patch manager
 *
 * @category DbPatcher
 * @package  Patcher
 */
class Patcher
{

    /**
     * @var PDO
     */
    private $connection;

    /**
     * @var string
     */
    private $path;

    /**
     * @var array
     */
    private $state = array();

    /**
     * @var string
     */
    private $stateFile;

    /**
     * Constructor, set patch path
     *
     * @param string $path Path to patch
     * @param PDO    $con  Database connection
     *
     * @return void
     */
    public function __construct($path, \PDO $con)
    {
        $this->path       = $path;
        $this->connection = $con;
        $this->stateFile  = $path . DIRECTORY_SEPARATOR . 'patcher_state.txt';

        if (file_exists($this->stateFile)) {
            $this->loadState();
        } else {
            touch($this->stateFile);
        }
    }

    /**
     * Return array as patch list
     * Key => patch's name
     * Value => boolean, Installed or not
     *
     * @return array
     */
    public function getList()
    {
        $patches = array();

        $iterator = new \DirectoryIterator($this->path);
        foreach ($iterator as $file) {
            if ($file->isFile() && preg_match('/^patch[0-9]{4}\.php$/', $file->getFilename())) {
                array_push($patches, str_replace('.php', '', $file->getFilename()));
            }
        }

        return $patches;
    }

    /**
     * Get database status
     *
     * @return array
     */
    public function status()
    {
        $status = array(
                   'status'    => 0,
                   'toInstall' => array(),
                  );

        $state = $this->state;
        if (!empty($state)) {
            $status['status'] = $this->getVersion(array_pop($state));
        }

        foreach ($this->getList() as $patch) {
            if (!$this->isInstalled($patch)) {
                array_push($status['toInstall'], $patch);
            }
        }

        return $status;
    }

    /**
     * Database upgrading
     *
     * @param int $version Target version, if NULL upgrade all patch
     *
     * @return bool
     */
    public function upgrade($version = null)
    {
        if (is_null($version)) {
            $patchTarget = array_pop($this->getList());
            $version     = $this->getVersion($patchTarget);
        } else {
            $patchTarget = $this->getNameFromVersion($version);
        }

        foreach ($this->getList() as $patchName) {
            if ($this->isInstalled($patchName)) {
                continue;
            }

            $patch = $this->loadPatch($patchName);
            $patch->upgrade();
            $patch->execute();

            array_push($this->state, $patchName);

            if ($patchName === $patchTarget) {
                break;
            }
        }

        $this->refreshStateFile();
    }

    /**
     * Database downgrading
     *
     * @param int $version Target version, if NULL downgrade all patch
     *
     * @return bool
     */
    public function downgrade($version = null)
    {
        if (is_null($version)) {
            $version = 0;
        }

        $patchTarget = $this->getNameFromVersion($version);
        $list        = array_reverse($this->getList());

        foreach ($list as $patchName) {
            if ($patchName === $patchTarget) {
                break;
            }

            if (!$this->isInstalled($patchName)) {
                continue;
            }

            $patch = $this->loadPatch($patchName);
            $patch->downgrade();
            $patch->execute();

            $stateKey = array_search($patchName, $this->state);
            unset($this->state[$stateKey]);
        }

        $this->refreshStateFile();
    }

    /**
     * Check if patch is already installed
     *
     * @param string $patch Patch to check
     *
     * @return bool
     */
    public function isInstalled($patch)
    {
        $this->checkPatchName($patch);
        return in_array($patch, $this->state);
    }

    /**
     * Load patcher state
     *
     * @return void
     */
    private function loadState()
    {
        $this->state = array();
        $state       = file($this->stateFile);

        foreach ($state as $patch) {
            $patch = trim($patch);

            if (empty($patch)) {
                continue;
            }

            $this->checkPatchName($patch);

            if ($this->patchExists($patch)) {
                array_push($this->state, $patch);
            }
        }
    }

    /**
     * Refresh state file
     *
     * @return bool
     */
    private function refreshStateFile()
    {
        $data = implode(PHP_EOL, $this->state);
        if (!file_put_contents($this->stateFile, $data)) {
            return false;
        }

        return true;
    }

    /**
     * Check if patch exists
     *
     * @param string $patch Patch to check
     *
     * @return bool
     */
    private function patchExists($patch)
    {
        if (file_exists($this->getPatchPath($patch))) {
            return true;
        }

        return false;
    }

    /**
     * Get patch's version
     *
     * @param string $patch Patch
     *
     * @return int
     */
    private function getVersion($patch)
    {
        $this->checkPatchName($patch);
        return (int) preg_replace('/patch[0]*/', '', $patch);
    }

    /**
     * Check is a valid patch name
     *
     * @param string $patch Patch to check
     *
     * @throws \UnexpectedValueException if name is invalid
     *
     * @return void
     */
    private function checkPatchName($patch)
    {
        if (!preg_match('/^patch[0-9]{4}$/', $patch)) {
            throw new \UnexpectedValueException('Patch name invalid (' . $patch . ')');
        }
    }

    /**
     * Get patch name from version
     *
     * @param int $version Version wanted
     *
     * @throws \LengthException If version too long
     *
     * @return string
     */
    private function getNameFromVersion($version)
    {
        $name  = 'patch';
        $count = strlen((string) $version);

        if ($count > 4) {
            throw new \LengthException('Version number accept only 4 characters');
        }

        $countZero = 4 - $count;

        for ($i = 0; $i < $countZero; $i++) {
            $name .= '0';
        }

        return $name . $version;
    }

    /**
     * Get patch's path
     *
     * @param string $patch Patch to get
     *
     * @return string
     */
    private function getPatchPath($patch)
    {
        return $this->path . DIRECTORY_SEPARATOR . $patch . '.php';
    }

    /**
     * Load a patch
     *
     * @param string $patch Patch to get
     *
     * @throws \InvalidArgumentException if patch not found
     *
     * @return Patch
     */
    private function loadPatch($patch)
    {
        $this->checkPatchName($patch);

        if (!$this->patchExists($patch)) {
            throw new \InvalidArgumentException('Patch not found');
        }

        $patchFile  = $this->getPatchPath($patch);
        $patchClass = ucfirst($patch);

        if (!class_exists($patchClass)) {
            include $patchFile;
        }

        return new $patchClass($this->connection);
    }
}
